zaza/openstack/utilities/upgrade_utils.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Collection of functions to support upgrade testing."""
+import re
+import logging
+import collections
+import zaza.model
+
+
+SERVICE_GROUPS = collections.OrderedDict([
+    ('Core Identity', ['keystone']),
+    ('Control Plane', [
+        'aodh', 'barbican', 'ceilometer', 'ceph-mon', 'ceph-fs',
+        'ceph-radosgw', 'cinder', 'designate',
+        'designate-bind', 'glance', 'gnocchi', 'heat', 'manila',
+        'manila-generic', 'neutron-api', 'neutron-gateway', 'placement',
+        'nova-cloud-controller', 'openstack-dashboard']),
+    ('Data Plane', [
+        'nova-compute', 'ceph-osd', 'swift-proxy', 'swift-storage'])
+])
+
+UPGRADE_EXCLUDE_LIST = ['rabbitmq-server', 'percona-cluster']
+
+
+def get_upgrade_candidates(model_name=None):
+    """Extract list of apps from model that can be upgraded.
+
+    :param model_name: Name of model to query.
+    :type model_name: str
+    :returns: List of application that can have their payload upgraded.
+    :rtype: []
+    """
+    status = zaza.model.get_status(model_name=model_name)
+    candidates = {}
+    for app, app_config in status.applications.items():
+        # Filter out subordinates
+        if app_config.get("subordinate-to"):
+            logging.warning(
+                "Excluding {} from upgrade, it is a subordinate".format(app))
+            continue
+
+        # Filter out charms on the naughty list
+        charm_name = extract_charm_name_from_url(app_config['charm'])
+        if app in UPGRADE_EXCLUDE_LIST or charm_name in UPGRADE_EXCLUDE_LIST:
```

Comment:
Maybe we need to move the handling of the exclude so get_upgrade_candidates returns everything and then have individual ones for series, openstack and charm upgrade which use get_upgrade_candidates to get the list and then they apply their exclude lists.